### PR TITLE
libspatialite: added patch to wrap call to xmlNanoHTTPCleanup()

### PR DESCRIPTION
**Pull request by b-rodrigues**
See https://github.com/NixOS/nixpkgs/issues/347085

Added patch to wrap the call to `xmlNanoHTTPCleanup()` as described here: https://www.gaia-gis.it/fossil/libspatialite/info/7c452740fe

This is enough to successfully build `rPackages.icosa`, but more generally it is likely that other packages could have issues due to re-enabling HTTP support in `libxml2` through an override in their buildInputs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by b-rodrigues** [comment]
`nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"` makes my PC shut itself down

**Comment by autra** [comment]
> nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD" makes my PC shut itself down

you might want to add `--build-args "--cores <n> --max-jobs  <m>"` where n*m is less than your number of cpu core minus a margin of your choosing to the nixpkgs-review command.

That being said, usually when one forgets that, the worst that can happen is that the memory and the swap are completely full, which will freeze your PC for some minutes before the linux oom killer decides to kill some stuff, but it should not shut down. If your pc shuts down, that is a bit worrisome. You should monitor temperature... Anyway, that's outside the scope of this PR.

**Comment by autra** [comment]
As for the PR itself, is there a patch upstream already? Considering how it affects a lot of geospatial packages in nix, it'd be nice to have an answer in the upstream issue before merging that, if we can wait...

File: pkgs/development/libraries/libspatialite/xmlNanoHTTPCleanup.patch
Comment (jtojnar) [inline pkgs/development/libraries/libspatialite/xmlNanoHTTPCleanup.patch:1]: You will probably also want to make conditional the header include.

File: pkgs/development/libraries/libspatialite/default.nix
Comment (jtojnar) [inline pkgs/development/libraries/libspatialite/default.nix:36]: It would be nice to move the comment inside so that it clearly matches the specific patch if more patches are added:

```nix
  patches = [
    # Drop use of deprecated libxml2 HTTP API.
    # From: https://www.gaia-gis.it/fossil/libspatialite/info/7c452740fe
    # see also: https://github.com/NixOS/nixpkgs/issues/347085
    ./xmlNanoHTTPCleanup.patch
  ];
```

Also, patches conventionally belong just after `src`.

File: pkgs/development/libraries/libspatialite/xmlNanoHTTPCleanup.patch
Comment (jtojnar) [inline pkgs/development/libraries/libspatialite/xmlNanoHTTPCleanup.patch:1]: Looking at the source, apparently the WFS is relying on libxml2 for fetching WFS over HTTP:

https://github.com/Dushistov/libspatialite/blob/156e902140c96dbcdc87409ac5824b1e6575ad84/src/wfs/wfs_in.c#L1294

Actually, all loading XML from web will be broken:

https://github.com/Dushistov/libspatialite/blob/156e902140c96dbcdc87409ac5824b1e6575ad84/src/gaiageo/gg_xml.c#L2891

I guess the project could use `xmlSetExternalEntityLoader` to specify custom fetcher (e.g. using curl) but that would not be trivial.

**Comment by kirillrdy** [comment]
Result of `nixpkgs-review pr 347724` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages marked as broken and skipped:</summary>
  <ul>
    <li>python311Packages.worldengine</li>
    <li>python311Packages.worldengine.dist</li>
    <li>python312Packages.shimmy</li>
    <li>python312Packages.shimmy.dist</li>
    <li>python312Packages.worldengine</li>
    <li>python312Packages.worldengine.dist</li>
    <li>worldengine-cli</li>
    <li>worldengine-cli.dist</li>
  </ul>
</details>
<details>
  <summary>11 packages failed to build:</summary>
  <ul>
    <li>python311Packages.cartopy</li>
    <li>python311Packages.cartopy.dist</li>
    <li>python311Packages.python-mapnik</li>
    <li>python311Packages.python-mapnik.dist</li>
    <li>python312Packages.cartopy</li>
    <li>python312Packages.cartopy.dist</li>
    <li>python312Packages.python-mapnik</li>
    <li>python312Packages.python-mapnik.dist</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
    <li>t-rex</li>
  </ul>
</details>
<details>
  <summary>146 packages built:</summary>
  <ul>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>cloudcompare</li>
    <li>entwine</li>
    <li>gdal (python312Packages.gdal)</li>
    <li>gdalMinimal</li>
    <li>gmt</li>
    <li>gplates</li>
    <li>grass</li>
    <li>haskellPackages.hgdal</li>
    <li>haskellPackages.hgdal.doc</li>
    <li>librasterlite2</li>
    <li>libspatialite</li>
    <li>libspatialite.dev</li>
    <li>mapcache</li>
    <li>mapnik</li>
    <li>mapproxy</li>
    <li>mapproxy.dist</li>
    <li>mapserver</li>
    <li>merkaartor</li>
    <li>mysql-workbench</li>
    <li>octavePackages.mapping</li>
    <li>openorienteering-mapper</li>
    <li>osmscout-server</li>
    <li>paraview</li>
    <li>pdal</li>
    <li>perl536Packages.Tirex</li>
    <li>perl536Packages.Tirex.devdoc</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>postgresql12JitPackages.postgis</li>
    <li>postgresql12JitPackages.postgis.doc</li>
    <li>postgresql12Packages.postgis</li>
    <li>postgresql12Packages.postgis.doc</li>
    <li>postgresql13JitPackages.postgis</li>
    <li>postgresql13JitPackages.postgis.doc</li>
    <li>postgresql13Packages.postgis</li>
    <li>postgresql13Packages.postgis.doc</li>
    <li>postgresql14JitPackages.postgis</li>
    <li>postgresql14JitPackages.postgis.doc</li>
    <li>postgresql14Packages.postgis</li>
    <li>postgresql14Packages.postgis.doc</li>
    <li>postgresql15JitPackages.postgis</li>
    <li>postgresql15JitPackages.postgis.doc</li>
    <li>postgresql15Packages.postgis</li>
    <li>postgresql15Packages.postgis.doc</li>
    <li>postgresqlJitPackages.postgis (postgresql16JitPackages.postgis)</li>
    <li>postgresqlJitPackages.postgis.doc (postgresql16JitPackages.postgis.doc)</li>
    <li>postgresqlPackages.postgis (postgresql16Packages.postgis)</li>
    <li>postgresqlPackages.postgis.doc (postgresql16Packages.postgis.doc)</li>
    <li>python311Packages.bsuite</li>
    <li>python311Packages.bsuite.dist</li>
    <li>python311Packages.django-bootstrap4</li>
    <li>python311Packages.django-bootstrap4.dist</li>
    <li>python311Packages.django-bootstrap5</li>
    <li>python311Packages.django-bootstrap5.dist</li>
    <li>python311Packages.fiona</li>
    <li>python311Packages.fiona.dist</li>
    <li>python311Packages.folium</li>
    <li>python311Packages.folium.dist</li>
    <li>python311Packages.gdal</li>
    <li>python311Packages.geodatasets</li>
    <li>python311Packages.geodatasets.dist</li>
    <li>python311Packages.geopandas</li>
    <li>python311Packages.geopandas.dist</li>
    <li>python311Packages.geoparquet</li>
    <li>python311Packages.geoparquet.dist</li>
    <li>python311Packages.inequality</li>
    <li>python311Packages.inequality.dist</li>
    <li>python311Packages.libpysal</li>
    <li>python311Packages.libpysal.dist</li>
    <li>python311Packages.mapclassify</li>
    <li>python311Packages.mapclassify.dist</li>
    <li>python311Packages.momepy</li>
    <li>python311Packages.momepy.dist</li>
    <li>python311Packages.morecantile</li>
    <li>python311Packages.morecantile.dist</li>
    <li>python311Packages.osmnx</li>
    <li>python311Packages.osmnx.dist</li>
    <li>python311Packages.plotnine</li>
    <li>python311Packages.plotnine.dist</li>
    <li>python311Packages.pygmt</li>
    <li>python311Packages.pygmt.dist</li>
    <li>python311Packages.pyogrio</li>
    <li>python311Packages.pyogrio.dist</li>
    <li>python311Packages.rasterio</li>
    <li>python311Packages.rasterio.dist</li>
    <li>python311Packages.rio-tiler</li>
    <li>python311Packages.rio-tiler.dist</li>
    <li>python311Packages.rioxarray</li>
    <li>python311Packages.rioxarray.dist</li>
    <li>python311Packages.shimmy</li>
    <li>python311Packages.shimmy.dist</li>
    <li>python311Packages.wktutils</li>
    <li>python311Packages.wktutils.dist</li>
    <li>python312Packages.bsuite</li>
    <li>python312Packages.bsuite.dist</li>
    <li>python312Packages.django-bootstrap4</li>
    <li>python312Packages.django-bootstrap4.dist</li>
    <li>python312Packages.django-bootstrap5</li>
    <li>python312Packages.django-bootstrap5.dist</li>
    <li>python312Packages.fiona</li>
    <li>python312Packages.fiona.dist</li>
    <li>python312Packages.folium</li>
    <li>python312Packages.folium.dist</li>
    <li>python312Packages.geodatasets</li>
    <li>python312Packages.geodatasets.dist</li>
    <li>python312Packages.geopandas</li>
    <li>python312Packages.geopandas.dist</li>
    <li>python312Packages.geoparquet</li>
    <li>python312Packages.geoparquet.dist</li>
    <li>python312Packages.inequality</li>
    <li>python312Packages.inequality.dist</li>
    <li>python312Packages.libpysal</li>
    <li>python312Packages.libpysal.dist</li>
    <li>python312Packages.mapclassify</li>
    <li>python312Packages.mapclassify.dist</li>
    <li>python312Packages.momepy</li>
    <li>python312Packages.momepy.dist</li>
    <li>python312Packages.morecantile</li>
    <li>python312Packages.morecantile.dist</li>
    <li>python312Packages.osmnx</li>
    <li>python312Packages.osmnx.dist</li>
    <li>python312Packages.plotnine</li>
    <li>python312Packages.plotnine.dist</li>
    <li>python312Packages.pygmt</li>
    <li>python312Packages.pygmt.dist</li>
    <li>python312Packages.pyogrio</li>
    <li>python312Packages.pyogrio.dist</li>
    <li>python312Packages.rasterio</li>
    <li>python312Packages.rasterio.dist</li>
    <li>python312Packages.rio-tiler</li>
    <li>python312Packages.rio-tiler.dist</li>
    <li>python312Packages.rioxarray</li>
    <li>python312Packages.rioxarray.dist</li>
    <li>python312Packages.wktutils</li>
    <li>python312Packages.wktutils.dist</li>
    <li>qgis</li>
    <li>qgis-ltr</li>
    <li>qmapshack</li>
    <li>saga</li>
    <li>spatialite_gui</li>
    <li>spatialite_tools</li>
    <li>sumo</li>
    <li>udig</li>
    <li>valhalla</li>
    <li>vpv</li>
  </ul>
</details>

**Comment by b-rodrigues** [comment]
> > nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD" makes my PC shut itself down
> 
> you might want to add `--build-args "--cores <n> --max-jobs <m>"` where n*m is less than your number of cpu core minus a margin of your choosing to the nixpkgs-review command.

Thanks!

> That being said, usually when one forgets that, the worst that can happen is that the memory and the swap are completely full, which will freeze your PC for some minutes before the linux oom killer decides to kill some stuff, but it should not shut down. If your pc shuts down, that is a bit worrisome. You should monitor temperature... Anyway, that's outside the scope of this PR.

I exaggerated for comedic effect: it just stays frozen forever and I need to manually shut it down.

File: pkgs/development/libraries/libspatialite/default.nix
Comment (autra) [reply pkgs/development/libraries/libspatialite/default.nix]: > Also, patches conventionally belong just after src.

Indeed, it's better! Do you know if there is a document somewhere describing these kinds of convention or is it just a usage thing?

**Comment by autra** [comment]
> I exaggerated for comedic effect: it just stays frozen forever and I need to manually shut it down.

aha, comedic effect successfull ;-) So my bet is that you've exhausted your memory and swap. Solution is to reduce the number of jobs * cores, so that the memory usage is also reduced.

File: pkgs/development/libraries/libspatialite/xmlNanoHTTPCleanup.patch
Comment (autra) [reply pkgs/development/libraries/libspatialite/xmlNanoHTTPCleanup.patch:1]: > Actually, all loading XML from web will be broken:
> 
> [Dushistov/libspatialite@156e902/src/gaiageo/gg_xml.c#L2891](https://github.com/Dushistov/libspatialite/blob/156e902140c96dbcdc87409ac5824b1e6575ad84/src/gaiageo/gg_xml.c#L2891)

This seems pretty impactful. I don't see how we can merge that without an upstream patch.

**Comment by b-rodrigues** [comment]
BTW with this patch, it also becomes possible to remove the override to `libxml2` in `gdal`

**Comment by b-rodrigues** [comment]
And also 

- `pdal`
- `gmt`
- `mysql-workbench`
- `grass`
- `mapserver`
- `librasterlite2`
- `mapnik`
- `postgresql`

as in, they build successfully at least.

Packages where this is not enough:

- `armagetronad` (looks like a fun game btw)

**Comment by autra** [comment]
> as in, they build successfully at least.

yeps, but I worry about the runtime consequences.

I've asked the question on the spatialite mailing list (pending approval), and also to the freebsd maintainer (they have the same issue of course) to see what they think.

File: pkgs/development/libraries/libspatialite/default.nix
Comment (jtojnar) [reply pkgs/development/libraries/libspatialite/default.nix]: I collected some of them here https://github.com/jtojnar/nixpkgs-hammering but mostly you need to get feel for conventions by osmosis.

**Comment by b-rodrigues** [comment]
> **Edit(jtojnar)**: It has been approved: https://groups.google.com/g/spatialite-users/c/FLBqJNIDkNQ

So, does that mean we could merge this in good conscience? What about the broken packages here 

> Result of `nixpkgs-review pr 347724` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 8 packages marked as broken and skipped:
> 11 packages failed to build:
> * python311Packages.cartopy
> * python311Packages.cartopy.dist
> * python311Packages.python-mapnik
> * python311Packages.python-mapnik.dist
> * python312Packages.cartopy
> * python312Packages.cartopy.dist
> * python312Packages.python-mapnik
> * python312Packages.python-mapnik.dist
> * pytrainer
> * pytrainer.dist
> * t-rex

**Comment by jtojnar** [comment]
> So, does that mean we could merge this in good conscience? What about the broken packages here

I was talking about the mailing list submission moderation. The runtime concerns still apply.

**Comment by b-rodrigues** [comment]
> I was talking about the mailing list submission moderation. The runtime concerns still apply.

Anything I could try to test this?

**Comment by nwellnhof** [comment]
> Edit(jtojnar): It has been approved: https://groups.google.com/g/spatialite-users/c/FLBqJNIDkNQ

The comment by Sandro is mostly wrong. This has nothing to do with memory leaks. The call to `xmlNanoHTTPCleanup` is only required to pick up a changes to the HTTP_PROXY environment variable in long-running processes.

> Anything I could try to test this?

You could try to use the `.loadwfs` command of the [spatiallite CLI](https://www.gaia-gis.it/fossil/libspatialite/wiki?name=dot+macros). This requires a WFS data source on an HTTP server. Note that HTTPS never was and never will be supported.

**Comment by b-rodrigues** [comment]
I'm trying the following example from the documentation:

```
.loadwfs http://webgis.regione.sardegna.it/geoserver/ows?service=WFS&request=GetFeature&typeName=dbu:ASL dbu:ASL asl_sardegna;
```

and it just hangs. Cancelling I get the following:

```
I/O warning : failed to load "http://webgis.regione.sardegna.it/geoserver/ows?service=WFS&request=GetFeature&typeName=dbu:ASL": No such file or directory
```

but it's the same from `master`, so I guess it really is down. I've tried with the other urls listed [here](https://www.gaia-gis.it/fossil/libspatialite/wiki?name=WFS), no luck.

**Comment by autra** [comment]
> The comment by Sandro is mostly wrong. This has nothing to do with memory leaks. The call to xmlNanoHTTPCleanup is only required to pick up a changes to the HTTP_PROXY environment variable in long-running processes.

@nwellnhof thanks a lot for chiming it, it is very very much appreciated (and save us a lot of time).

So honestly, I don't think we should care much about the lack of update of HTTP_PROXY in a long running process (which should be already the exception with sqlite), for a feature that is probably not useful any more (as it never supported https anyway).

The code looks fine, and will probably help for other packages, including at least qgis. Thanks @b-rodrigues for your patience!

**Review by autra — APPROVED** [review]

**Comment by b-rodrigues** [comment]
A new R release is planned for the 31st of October, meaning we will also be bumping the rPackages tree; it'd be great if this patch could get merged by then, as it would mean we could get the rPackages tree back to a quite healthy state.

**Comment by imincik** [comment]
Thanks everybody for investigation. Can we finally merge this PR ?

**Comment by autra** [comment]
From my point of view, yes, we should merge that as soon as possible yes. We might have some cleaning to do in the geospatial owned packages after (removing the `withHttp = true` thing in packages that don't really need it), but it can probably wait.

**Comment by imincik** [comment]
> From my point of view, yes, we should merge that as soon as possible yes. We might have some cleaning to do in the geospatial owned packages after (removing the `withHttp = true` thing in packages that don't really need it), but it can probably wait.

Yes, once this PR is merged we should start cleaning our package from `withHttp = true` workaround.

**Review by imincik — APPROVED** [review]

**Pull request merged**
